apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -0,0 +1,248 @@
+
+  Skills possess effects that affect other units' or oneself's stats (the on defined in the effect's `stat_affected` field).
+
+  They have different application types (checked are implemented):
+  [x] Instant - Applied once, irreversible.
+  [ ] Permanent - Applied once, is stored in the unit so that it can be reversed (with a dispel, for example)
+  [ ] Duration - Applied once and reverted once its duration ends.
+  [ ] Periodic - Applied many times every x number of steps for a total duration of y steps
+
+  They also have different targeting strategies:
+  [x] Random
+  [ ] Nearest
+  [ ] Furthest
+  [ ] Min Health
+  [ ] Max Health
+  [ ] Min Shield
+  [ ] Max Shield
+  [ ] Frontline - Heroes in slots 1 and 2
+  [ ] Backline - Heroes in slots 2 to 4
+  [ ] Factions
+  [ ] Classes
+
+  Two units can attack the same unit at the same time and over-kill them. This is expected behavior that results
+  from having the battle be simultaneous.
+
+  """
+  alias GameBackend.Units.Skills.Effect
+  alias GameBackend.Units.Skill
+  alias Champions.Units
+  alias GameBackend.Units.Unit
+
+  require Logger
+
+  @maximum_steps 500
+
+  @doc """
+  Runs a battle between two teams.
+  Teams are expected to be lists of units with their character and their skills preloaded.
+
+  Returns `:team_1`, `:team_2`, `:tie` or `:timeout`.
+
+  ## Examples
+
+      iex> team_1 = Enum.map(user1.units, GameBackend.Repo.preload([character: [:basic_skill, :ultimate_skill]]))
+      iex> team_2 = Enum.map(user2.units, GameBackend.Repo.preload([character: [:basic_skill, :ultimate_skill]]))
+      iex> run_battle(team_1, team_2)
+      :team_1
+  """
+  def run_battle(team_1, team_2, seed \\ 1) do
+    :rand.seed(:default, seed)
+    team_1 = Enum.into(team_1, %{}, fn unit -> create_unit_map(unit, 1) end)
+    team_2 = Enum.into(team_2, %{}, fn unit -> create_unit_map(unit, 2) end)
+    units = Map.merge(team_1, team_2)
+
+    # The initial_step_state is what allows the battle to be simultaneous. If we refreshed the accum on every action,
+    # we would be left with a turn-based battle. Instead we take decisions based on the state of the battle at the beggining
+    # of the step regardless of the changes that happened "before" (execution-wise) in this step.
+    Enum.reduce_while(1..@maximum_steps, units, fn step, initial_step_state ->
+      new_state =
+        initial_step_state
+        |> Enum.map(fn {id, _unit} -> id end)
+        |> Enum.take_random(Enum.count(units))
+        |> Enum.reduce(initial_step_state, fn unit_id, current_state ->
+          process_step(initial_step_state[unit_id], initial_step_state, current_state)
+        end)
+
+      remove_dead_units(new_state)
+      |> check_winner(step)
+    end)
+  end
+
+  defp process_step(unit, initial_step_state, current_state) do
+    new_state =
+      cond do
+        not can_attack(unit) ->
+          current_state
+
+        can_cast_ultimate(unit) ->
```

Comment:
What do you think of naming it `can_cast_ultimate_skill`?
Same for `can_cast_basic`.